Use strlen when testing validity of API string argument length

cgolubi1 opened the pull request:
* Fixes #3030 

The new responder test checks provide evidence that the fix works.  Without the fix, a similar test shows that game creation succeeds, but the description is corrupted:

```
1) responder00Test::test_request_createGame
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'αααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααα'
+'ααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααααα?'

/buttonmen/test/src/api/responder00Test.php:345

FAILURES!
Tests: 2, Assertions: 74, Failures: 1.
```

I also tested manually and found that with the fix, 127 alphas succeed and 128 alphas fail cleanly.  Without the fix, i see a variety of unhelpful behavior as described in comments to #3030

blackshadowshade commented:
I'm fine with the fix per se. I just wonder whether we should rename 'maxlength' to 'maxsize' to make it clear that we're talking about a number of bytes instead of a number of characters.

blackshadowshade commented:
(I'm not going to block the pull request based on that request if you think it's a bad idea)

cgolubi1 commented:
I guess i would say it doesn't super matter --- this is an internal function used by the API specification; it's not a player-visible part of the API.

That said, i'm happy to change the variables in ApiSpec to maxbytes and minbytes, to make that clear.

Worth flagging: this new code is definitely *not* consistent with the length validation the UI does --- and i'm not all that it was consistent before this update either.  So there are UI uses of maxlength that we probably want to revisit, so we can give UI players a friendlier error.

However, i don't think we should futz with that now; i think we should get this PR over the fence, because i'd rather have players see "Argument (description) to function createGame was invalid", than be able to create games which block their overview page from loading, which is a live bug we have right now.

cgolubi1 commented:
Update: i looked at ApiSpec a bit, and sorry to be a coward about it, but the problem is that we're using `maxlength` two different ways: one is maximum length of a string (the case we changed here to make it correctly Unicode-friendly), and one is maximum length of an array (i.e. maximum number of elements).

So it's not just a question of copy/paste --- if i changed all the strings to `maxbytes`, i would have to carefully parse out which ones are the strings and which ones are the arrays, and that would make the PR more error-prone, at which point it would definitely need dev testing.

I guess i could change everything to `maxsize`, but i don't like `maxsize`; i don't think it's especially clearer than `maxlength`:
* Argument against "size": there's an implicit "in what unit" that we're not answering in the variable name
* Argument for "length": after all, we're using the function `strlen()` [length] to get the number of bytes of the string, so there's a semantic argument that "length" is correct

In conclusion: philosophically, i like the idea of using "bytes" for strings and "entries" or "elements" for arrays, but i don't think it's worth making this PR a lot more complicated to approve for that small benefit.  Someone is going to have to return to this topic in the future to make a player-visible improvement to the UI length measurements, and they can think about these variable names in `ApiSpec` when they do so.

blackshadowshade commented:
Fair enough, we'll deal with this in the future if we need to then.
